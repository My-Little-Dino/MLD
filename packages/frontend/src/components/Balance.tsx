import React, { useEffect } from 'react';
import styled from 'styled-components';
import { useWeb3React } from '@web3-react/core';
import { toBech32 } from '@harmony-js/crypto'

function Balance() {
	const { account, library, chainId } = useWeb3React()
	const isHmyLibrary = (library?.messenger?.chainType === 'hmy')
  
	const [balance, setBalance] = React.useState()
	React.useEffect((): any => {
	  if (!!account && !!library) {
		let stale = false
		const accountArgs = isHmyLibrary ? { address: toBech32(account) } : account
  
		library
		  .getBalance(accountArgs)
		  .then((balance: any) => {
			if (isHmyLibrary) {
			  balance = balance.result;
			}
			if (!stale) {
			  balance =  (balance / 1000000000000000000).toFixed(2);
			  setBalance(balance)
			}
		  })
		  .catch(() => {
			if (!stale) {
			  setBalance(undefined)
			}
		  })
  
		return () => {
		  stale = true
		  setBalance(undefined)
		}
	  }
	}, [account, library, chainId]) // ensures refresh if referential identity of library doesn't change across chainIds
  
	return (
	  <BalanceComponent>
		<span>Balance</span>
		<span role="img" aria-label="gold">
		  ðŸ’°
		</span>
		<span>{balance === null ? 'Error' : balance ? (isHmyLibrary ? balance : `${balance} ONE`) : ''}</span>
	  </BalanceComponent>
	)
  }

const BalanceComponent = styled.div`
	padding: 10px 20px;
	border-radius: 25px;
	background-color: white;
	color: black;

	& > span {
		margin-left: 4px;
	}
`;

export default Balance;
