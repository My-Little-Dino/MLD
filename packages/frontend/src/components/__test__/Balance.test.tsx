import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';

import { useWeb3React } from '@web3-react/core';
import { useHarmony } from '../../context/harmonyContext';

import Balance from '../Balance';

jest.mock('@web3-react/core');

jest.mock('../../context/harmonyContext');

describe('Test Balance component', () => {
	const component = <Balance />;

	const setup = () => render(component);

	beforeEach(() => {
		(useWeb3React as jest.Mock).mockReturnValue({
			account: '0xcfffa6ace98eb3464f77af3059389f8f2c18b6e9',
			chainId: '2',
		});
		(useHarmony as jest.Mock).mockReturnValue({
			balance: '0',
			fetchBalance: jest.fn(),
			resetBalance: jest.fn(),
		});
	});

	afterEach(jest.clearAllMocks);

	it('Should render component without crash', () => {
		const div = document.createElement('div');
		ReactDOM.render(component, div);
		ReactDOM.unmountComponentAtNode(div);
	});

	it('Should match snapshot', () => {
		const { container } = setup();
		expect(container).toMatchSnapshot();
	});

	it('Should call fetch balance if we have an account', () => {
		setup();
		expect(useHarmony().fetchBalance).toHaveBeenCalled();
	});

	it('Should call reset balance if we dont have an account', () => {
		(useWeb3React as jest.Mock).mockReturnValue({
			account: null,
			chainId: '2',
		});
		setup();
		expect(useHarmony().resetBalance).toHaveBeenCalled();
	});
});
