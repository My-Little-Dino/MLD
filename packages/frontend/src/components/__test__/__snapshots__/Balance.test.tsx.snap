// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Balance component Should match snapshot 1`] = `
<div>
  <div
    class="sc-bdnxRM fwNNIJ"
  >
    <b>
      0
    </b>
     
    <span>
      ONE
    </span>
  </div>
</div>
`;
